src/Microsoft.DotNet.VersionTools/Dependencies/Submodule/IndicatorPackageSubmoduleUpdater.cs

```diff
@@ -0,0 +1,96 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.VersionTools.Dependencies.Submodule
+{
+    /// <summary>
+    /// Downloads a package specified from a dependency build info and updates the target git
+    /// submodule to match the version inside.
+    /// </summary>
+    public class IndicatorPackageSubmoduleUpdater : SubmoduleUpdater
+    {
+        /// <summary>
+        /// The NuGet v2 base url to use to download the indicator package, without a trailing '/'.
+        /// For example, https://dotnet.myget.org/F/dotnet-core/api/v2/package.
+        /// </summary>
+        public string PackageDownloadBaseUrl { get; set; }
+
+        public string IndicatorPackageId { get; }
+
+        public IndicatorPackageSubmoduleUpdater(string indicatorPackageId)
+        {
+            if (indicatorPackageId == null)
+            {
+                throw new ArgumentNullException(nameof(indicatorPackageId), "An indicator package must be specified.");
+            }
+            IndicatorPackageId = indicatorPackageId;
+        }
+
+        protected override string GetDesiredCommitHash(
+            IEnumerable<DependencyBuildInfo> dependencyBuildInfos,
+            out IEnumerable<DependencyBuildInfo> usedBuildInfos)
+        {
+            foreach (var info in dependencyBuildInfos)
+            {
+                PackageIdentity package = info.Packages
+                    .FirstOrDefault(p => p.Id == IndicatorPackageId);
+
+                if (package == null)
+                {
+                    continue;
+                }
+
+                using (ZipArchive archive = DownloadPackageAsync(info, package).Result)
+                {
+                    Stream versionTxt = archive.GetEntry("version.txt")?.Open();
+                    if (versionTxt == null)
+                    {
+                        Trace.TraceWarning(
+                            $"Downloaded '{package}' in '{info.BuildInfo.Name}' " +
+                            $"to upgrade '{Path}', but it had no version.txt file. Skipping.");
+                        continue;
+                    }
+
+                    string packageCommitHash = new StreamReader(versionTxt).ReadLine();
+                    Trace.TraceInformation($"Found commit '{packageCommitHash}' in versions.txt.");
+
+                    usedBuildInfos = new[] { info };
+                    return packageCommitHash;
+                }
+            }
+
+            Trace.TraceError($"Failed to find '{IndicatorPackageId}' specifying a commit in any build-info.");
+            usedBuildInfos = Enumerable.Empty<DependencyBuildInfo>();
+            return null;
+        }
+
+        protected async Task<ZipArchive> DownloadPackageAsync(DependencyBuildInfo info, PackageIdentity package)
+        {
+            if (PackageDownloadBaseUrl == null)
```

Comment:
There's a plan to eventually have dotnet/versions repo list where the build published its packages, so for any repos that do that, this argument doesn't need to be set. (Then the `NotSupportedException` would be moved to the case where the property is null *and* the build-info doesn't list a download source.)